### PR TITLE
Signup: Adding segments API results to the data layer

### PR DESCRIPTION
**Pull request by ramonjd**
## Changes proposed in this Pull Request

We store some [site segment data](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/site-type.js) in Calypso. 

As always, we had good intentions, however in the meantime, the id values have converged. 😱 

We need to be in sync with the backend.

This PR adds the segments HTTP results to the state tree (we'll add the files to the bundle when we start using them in a subsequent PR)

Required by: https://github.com/Automattic/wp-calypso/pull/33087

## Testing instructions

`npm run test-client client/state/signup/segments`

`npm run test-client client/state/data-layer/wpcom/signup/segments`

### CODE REVIEW

File: client/state/data-layer/wpcom/signup/segments/test/index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+					apiNamespace: 'wpcom/v2',
+					method: 'GET',
+					path: '/segments',
+				},
+				mockAction
+			)
+		);
+	} );
+
+	test( 'storeSegments()', () => {
```
Comment (spen) [inline client/state/data-layer/wpcom/signup/segments/test/index.js]: I'm personally of the camp where this should be a `describe` block:
```js
describe( 'storeSegments()', () => {
```
And the `test` block should tell us what to expect from the test below:

```js
describe( 'storeSegments()', () => {
  test( 'should return the same output as setSegments', () => {
  ...
```
A pattern I've been playing with a little lately is to do all of the test action in a describe block and use a new test block to state each thing that we expect from the results (Using the next test as an example):

```js
describe( 'showVerticalsRequestError()', () => {
  const result = showSegmentsRequestError();

  test( 'should have a `type` attribute with value `NOTICE_CREATE`', () => {
  	expect( result.type ).toEqual( NOTICE_CREATE );	
  } );

  test( 'should have a notice attribute of plain object type', () => {
  	expect( isPlainObject( result.notice ) ).toEqual( true );
  } );

  test( 'should have a status attribute, on the notice object, with value `is-error`', () => {
  	expect( result.notice.status ).toEqual( 'is-error' );
  } );
} );
```
In this example the test set up is very slim, so it's hard to see the benefit of doing it this way, but you might imagine something like this (slightly contrived) example:

```js
describe( 'showVerticalsRequestError()', () => {
  const importerId = '12345';
  const importStatus = { id: importerId, type: 'someType' },
  const action = {
    importerId, 
    importStatus,
    type: IMPORTS_IMPORT_RECIEVE,
  };
  const existingImporterId = '54321';
  const existingImportStatus = { id: existingImporterId, type: 'someType' },
  const state = {
    [ existingImporterId ]: existingImportStatus,
  };
  const result = reducer( state, action );

  test( 'should add an importer item keyed by the given id, with a value of the given importerStatus', () => {
  	expect( result[ importerId ] ).toEqual( importStatus );	
  } );

  test( 'it does not affect existing importer items', () => {
  	expect( result[ existingImporterId ] ).toEqual( existingImporterStatus );	
  } );

```

File: client/state/signup/segments/actions.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import { SIGNUP_SEGMENTS_REQUEST, SIGNUP_SEGMENTS_SET } from 'state/action-types';
+
+import 'state/data-layer/wpcom/signup/segments';
+
+/**
+ * Action creator: Request segments data.
+ *
+ * @return {Object} The action object.
+ */
+export const requestSegments = () => ( {
```
Comment (spen) [inline client/state/signup/segments/actions.js]: As `requestSegments` is not dynamic, could we just return a plain object?
```js
export const requestSegments = {
  type: SIGNUP_SEGMENTS_REQUEST,
};
```

Comment (ramonjd) [reply client/state/signup/segments/actions.js]: I don't grok what you mean here. Could you elaborate?

Right now it's a redux action creator and I plan to use the action creator in a connect function, but if we returned an object we'd still have to dispatch the action somewhere, right?

File: client/state/data-layer/wpcom/signup/segments/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	http(
+		{
+			apiNamespace: 'wpcom/v2',
+			method: 'GET',
+			path: '/segments',
+		},
+		action
+	);
+
+export const storeSegments = segments => setSegments( segments );
```
[the remark this answers is not in the document]
Comment (ramonjd) [reply client/state/data-layer/wpcom/signup/segments/index.js]: > Do we anticipate storeSegments growing in logic?

Not any time soon. The only reason we're exporting here is for unit testing. It's a bit of an overkill, but 

You've actually pointed me to an error in my syntax. I've updated this in the next commit, which should hopefully address these comments.

`onSuccess` is eventually [dispatched](https://github.com/Automattic/wp-calypso/blob/b383328/client/state/data-layer/wpcom-http/utils.js#L187), and becomes wrapped in some magic middleware, but eventually we get back **two** arguments: a data-layer-related action and the data from the response, which we then pass to the `segments` action.

File: client/state/data-layer/wpcom/signup/segments/test/index.js (hunk 1 above)
Comment (ramonjd) [reply client/state/data-layer/wpcom/signup/segments/test/index.js]: > I'm personally of the camp where this should be a describe block:

Comment (ramonjd) [reply client/state/data-layer/wpcom/signup/segments/test/index.js]: In fact, I'm going to delete this test file as there are no real side effects. 

We're already testing our state methods, and the data layer is very well tested already.

File: client/state/data-layer/wpcom/signup/segments/index.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		action
+	);
+
+export const storeSegments = ( action, data ) => setSegments( data );
+export const showSegmentsRequestError = () =>
+	errorNotice(
+		translate( 'We encountered an error on fetching data from our server. Please try again.' )
+	);
+
+registerHandlers( 'state/data-layer/wpcom/signup/verticals', {
```
Comment (adamziel) [inline client/state/data-layer/wpcom/signup/segments/index.js]: I think this should say `segments` and not `verticals` so I renamed it. Feel free to rollback if it's wrong!

File: client/state/signup/segments/actions.js (hunk 2 above)
Comment (spen) [reply client/state/signup/segments/actions.js]: Sorry yeah! I think I was actually wrong here!
We can dispatch plain object actions like so:
```js
const someThunk = payload => dispatch => {
  dispatch( { type: 'SOME_ACTION, payload } );
}
```
So I thought we could just do that in this case, given that the resulting action doesn't change with variables etc. but I tested this and t doesn't work when importing and binding within `connect`s `mapDispatchToProps` argument.

So ignore this comment - I was barking up a tree that turned out to be wrong :P

File: client/state/data-layer/wpcom/signup/segments/test/index.js (hunk 1 above)
Comment (spen) [reply client/state/data-layer/wpcom/signup/segments/test/index.js]: > But you speak my linguine.

I'm glad we see carb to carb :)

File: client/state/signup/segments/actions.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+import { SIGNUP_SEGMENTS_REQUEST, SIGNUP_SEGMENTS_SET } from 'state/action-types';
+
+import 'state/data-layer/wpcom/signup/segments';
```
Comment (spen) [inline client/state/signup/segments/actions.js]: I just want to check, we import this file to trigger the calling of `registerHandlers( ... )` right?

Comment (ramonjd) [reply client/state/signup/segments/actions.js]: Just so we can bundle them when we need them. See: https://github.com/Automattic/wp-calypso/commit/e80cda04df36df1cc0bf7b7d2716920d5db00948

Comment (spen) [reply client/state/signup/segments/actions.js]: Nice! Thanks @ramonjd :)

**Pull request merged**